### PR TITLE
230 customizing results output

### LINKED ISSUE #230: customizing results output
from a user on Slack:

> As a data analyst doing ad-hoc SQL analysis I often times find myself creating SQL results in a notebook which contain URL values for a given column -- a bit of a "df as a dashboard" .
> Is there any easy way to enable these URLs in the result set to be clickable?  Maybe via a SqlMagic config?
> Currently I have a bit of a hack to do this by (1) returning the query results to a df, then (2) calling method to style it if it looks like a URI (see this notebook), but a simpler way seems desirable and a feature others may find useful as well.

https://gist.github.com/gregrahn/2a1196d6a6588a7c04c52b5970895937
https://github.com/softhints/Pandas-Tutorials/blob/master/styling/create-clickable-link-pandas-dataframe-jupyterlab.ipynb

I think we have two options: either we provide a style API like pandas do or we parse the results automatically.

I think it's fine to parse them automatically; at the beginning, I was concerned about performance but users only display small tables in the results. Alternatively, we might setup a threshold: if the table has more than X cells, do not apply the styling.

### PR DESCRIPTION
**Pull request by AnirudhVIyer**
## Issue ticket number and link
Closes #230 

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [x] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--274.org.readthedocs.build/en/274/

### CODE REVIEW

**Comment by edublancas** [comment]
is this ready for review? looks like you ran some markdown formatted cause the CHANGELOG was edited

**Comment by AnirudhVIyer** [comment]
Yeah, it is ready for review.

**Review by edublancas — CHANGES_REQUESTED** [review]
please check your formatting settings, I see a lot of changes in files that should not be edited (example, the CHANGELOG and some test files)

**Comment by AnirudhVIyer** [comment]
Will do that.

**Comment by AnirudhVIyer** [comment]
I have added a new table with urls in the conftest file to run a few tests. Is it alright?

**Comment by AnirudhVIyer** [comment]
Hi @edublancas can you approve the workflow checks. Thank You.

File: src/sql/run.py [hunk 1]
```diff
@@ -170,7 +173,28 @@ def DataFrame(self, payload):
         "Returns a Pandas DataFrame instance built from the result set."
         import pandas as pd
 
+        "create a method to check for urls"
```
Comment (edublancas) [inline src/sql/run.py:176]: this should be a comment, not a string

```suggestion
        # create a method to check for urls"
```

please also apply it to the other comments

**Review by edublancas — CHANGES_REQUESTED** [review]
overall it looks good, just added one minor comment.

I had an issue reproducing it for pandas data frames, am I doing something wrong? 

image

**Comment by AnirudhVIyer** [comment]
> overall it looks good, just added one minor comment.
> 
> I had an issue reproducing it for pandas data frames, am I doing something wrong?
> 
> image

Hi Eduardo, to make the URLs clickable in a dataframe, we need to add a style function that returns a Styler object, not the original dataframe. If we return the Styler object, users won't be able to conduct further analysis on it. Therefore, I added a styling function to the returned dataframe as an attribute. Users can use this attribute to display clickable links when they want to but still retain the original dataframe for further use.

you can display the dataframe with styling like this: 

```
df.styled_func()

```

File: src/sql/run.py (hunk 1 above)
Comment (AnirudhVIyer) [reply src/sql/run.py]: Made the changes

**Review by edublancas — CHANGES_REQUESTED** [review]
ok, since getting this to work with pandas involves users knowing a bit of the implementation details, I think it's better to only keep the URL parsing the ResultSet - it's ok if the links go away when turning it into a pandas data frame

**Comment by AnirudhVIyer** [comment]
> ok, since getting this to work with pandas involves users knowing a bit of the implementation details, I think it's better to only keep the URL parsing the ResultSet - it's ok if the links go away when turning it into a pandas data frame

Sure, I will make the required changes.

**Comment by AnirudhVIyer** [comment]
I have removed the pandas url attribute.

**Review by edublancas — APPROVED** [review]

**Comment by edublancas** [comment]
great work!

**Comment by AnirudhVIyer** [comment]
thank you!

**Pull request merged**